include/GFp/versioned.h

```diff
@@ -0,0 +1,400 @@
+#define GFp_sha256_block_data_order_shaext VERSIONED(GFp_sha256_block_data_order_shaext)
+#define GFp_sha256_block_data_order_ssse3 VERSIONED(GFp_sha256_block_data_order_ssse3)
+#define GFp_sha512_block_data_order VERSIONED(GFp_sha512_block_data_order)
+#define GFp_sha512_block_data_order_avx VERSIONED(GFp_sha512_block_data_order_avx)
+#define GFp_sha512_block_data_order_avx2 VERSIONED(GFp_sha512_block_data_order_avx2)
+#define GFp_size_t_align VERSIONED(GFp_size_t_align)
+#define GFp_size_t_size VERSIONED(GFp_size_t_size)
+#define GFp_uint16_t_align VERSIONED(GFp_uint16_t_align)
+#define GFp_uint16_t_size VERSIONED(GFp_uint16_t_size)
+#define GFp_uint32_t_align VERSIONED(GFp_uint32_t_align)
+#define GFp_uint32_t_size VERSIONED(GFp_uint32_t_size)
+#define GFp_uint64_t_align VERSIONED(GFp_uint64_t_align)
+#define GFp_uint64_t_size VERSIONED(GFp_uint64_t_size)
+#define GFp_uint8_t_align VERSIONED(GFp_uint8_t_align)
+#define GFp_uint8_t_size VERSIONED(GFp_uint8_t_size)
+#define GFp_uint_align VERSIONED(GFp_uint_align)
+#define GFp_uint_size VERSIONED(GFp_uint_size)
+#define GFp_vpaes_encrypt VERSIONED(GFp_vpaes_encrypt)
+#define GFp_vpaes_set_encrypt_key VERSIONED(GFp_vpaes_set_encrypt_key)
+#define GFp_x25519_NEON VERSIONED(GFp_x25519_NEON)
+#define GFp_x25519_ge_frombytes_vartime VERSIONED(GFp_x25519_ge_frombytes_vartime)
+#define GFp_x25519_ge_scalarmult_base VERSIONED(GFp_x25519_ge_scalarmult_base)
+#define GFp_x25519_public_from_private VERSIONED(GFp_x25519_public_from_private)
+#define GFp_x25519_sc_muladd VERSIONED(GFp_x25519_sc_muladd)
+#define GFp_x25519_sc_reduce VERSIONED(GFp_x25519_sc_reduce)
+#define GFp_x25519_scalar_mult VERSIONED(GFp_x25519_scalar_mult)
+#define GFp_x25519_x86_64 VERSIONED(GFp_x25519_x86_64)
+#define GFp_x25519_x86_64_freeze VERSIONED(GFp_x25519_x86_64_freeze)
+#define GFp_x25519_x86_64_ladderstep VERSIONED(GFp_x25519_x86_64_ladderstep)
+#define GFp_x25519_x86_64_mul VERSIONED(GFp_x25519_x86_64_mul)
+#define GFp_x25519_x86_64_square VERSIONED(GFp_x25519_x86_64_square)
+#define GFp_x25519_x86_64_work_cswap VERSIONED(GFp_x25519_x86_64_work_cswap)
+
+// LIMBS symbols.
+#define LIMBS_add_assign VERSIONED(LIMBS_add_assign)
+#define LIMBS_add_mod VERSIONED(LIMBS_add_mod)
+#define LIMBS_are_zero VERSIONED(LIMBS_are_zero)
+#define LIMBS_equal VERSIONED(LIMBS_equal)
+#define LIMBS_less_than VERSIONED(LIMBS_less_than)
+#define LIMBS_reduce_once VERSIONED(LIMBS_reduce_once)
+#define LIMBS_shl_mod VERSIONED(LIMBS_shl_mod)
+#define LIMBS_sub_assign VERSIONED(LIMBS_sub_assign)
+#define LIMBS_sub_mod VERSIONED(LIMBS_sub_mod)
+#define LIMBS_sub_mod_ex VERSIONED(LIMBS_sub_mod_ex)
+
+// ??
+#define bssl_constant_time_test_main VERSIONED(bssl_constant_time_test_main)
+#define gfp_little_endian_bytes_from_scalar VERSIONED(gfp_little_endian_bytes_from_scalar)
+
+// For MacOS symbols generated from perl scripts.
+#define _GFp_AES_encrypt VERSIONED(_GFp_AES_encrypt)
```

Comment:
When we're on macOS, we need both versions. The first will be used by C regardless while the second will be used by the generated assembly code.

We could, however, generate the second set on macOS only. IE, we'd kill all of the `_` ones from `versioned.h`. At build time, check if we're on macOS and emit a new `versioned.h` with the `_` symbols. If we're not on macOS, we'd simply copy the file over.

Reply:
> The first will be used by C regardless while the second will be used by the generated assembly code.

Sure. But we can make it so that `GFp_VERSIONED()` expands to a name with a leading underscore only when it is included from an assembly file, and expands without the leading underscore all the rest of the time. Then we can completely eliminate the duplication.

Reply:
No, that's not correct. The `#define` has to redefine the underscored symbol (`_sym`) to a versioned underscored symbol (`_sym_vxxx`) since `_sym` is what's in the assembly file. We absolutely need something that's effectively:

```c
#define _sym VERSIONED(_sym)
```